hvantk/core/utils/hail_context.py

```
@@ -83,6 +83,29 @@ def init_hail(**kwargs) -> None:
                     kwargs,
                 )
             return
+        # A Hail backend may already exist from a path other than init_hail()
+        # — e.g. a raw hl.init() in a notebook or a test fixture. Calling
+        # hl.init() again raises Hail's "already initialized" error, so adopt
+        # the existing session instead of re-initializing.
+        try:
+            hl.current_backend()
+            already_running = True
+        except Exception:
```

Ruff (0.15.15)

[warning] Do not catch blind exception: `Exception`

(BLE001)